New package to auto-generate public API documentation

oandregal opened the pull request:
This PR introduces a new package called `docgen` to help us generate the _public API_ of our code. Given an entry point file, it outputs the existing ES6 export statements and their corresponding JSDoc comments in human-readable format.

## Usage

> npx docgen <relative-path-to-entry-point.js>

This command will generate a file named `entry-point-api.md` containing all the exports and its JSDoc comments.

**Example:**

Let's use it with the `wordcount` package:

> npx docgen packages/wordcount/src/index.js

and then inspect the output

> less packages/wordcount/src/index-api.md

Find more details, documentation, and examples in the [README.md](https://github.com/WordPress/gutenberg/blob/try/api-docs/packages/docgen/README.md)

## Testing

Play with this to create documentation for a Gutenberg package, components, etc and report back!

## Coda

Note that this is only intended to work with native ES6 modules, not CommonJS. We have some packages used as config files or hooking into other tools (webpack, babel, etc) that are unsupported. See [coverage.md](packages/docgen/coverage.md).

oandregal committed:
First, find export declarations

oandregal committed:
Second, parse JSDoc

oandregal committed:
Third, create markdown files with exports and JDoc descriptions.

oandregal committed:
Refactor and prevent against no JSDoc

oandregal committed:
Move script to own package

oandregal committed:
Split api into cli, engine, and formatter

oandregal committed:
Add initial tests

oandregal committed:
Test named export (const)

oandregal committed:
Add test for default named functions

oandregal committed:
Add support for default export of anonymous functions

oandregal committed:
Better phrasing for test

oandregal committed:
Add support for default exports (identifiers)

oandregal committed:
Add support for named exports of a single identifier

oandregal committed:
Add test for many exports at once

oandregal committed:
Tweak text and reorder tests

chrisvanpatten commented:
It would be helpful to clarify how this relates to the work that @herregroen has done in getting jsdoc parsing for the core DevHub :)

oandregal committed:
Teach engine to take JSDoc from identifier declaration

oandregal committed:
Cover undocumented exports

oandregal committed:
Be more explicit about what is a test file

oandregal committed:
Add more test to cover potential regressions

oandregal committed:
Expose script as npm script

oandregal committed:
CLI: manage errors

oandregal committed:
Refactor to pave the way to next step

oandregal committed:
Refactor to prepare next step

oandregal committed:
Refactor to prepare next step

oandregal committed:
Add helper functions

oandregal committed:
Add mechanism to query upper layers for dependency code

oandregal committed:
Engine can lookup JSDoc in dependency file

oandregal committed:
Formatter - adapt a new engine API

oandregal committed:
Add test for looking up JSDoc in dependency

oandregal committed:
CLI: adapt to new engine API

oandregal committed:
Engine: enable JSX support

oandregal committed:
Evaluate how much coverage do we have already

oandregal committed:
Remove unused file

oandregal committed:
Add test for multiple named exports

oandregal committed:
Add more info for debugging purposes

oandregal committed:
Add tests for get name of named function and variable

oandregal committed:
Add support and test for parsing classes

oandregal committed:
Add test for single identifier

oandregal committed:
Add test for multiple named identifiers

oandregal committed:
Add tests for default export of functions (anonymous and named)

oandregal committed:
Add tests for default export of variables

oandregal committed:
Add tests for default export of classes

oandregal committed:
Add tests for default export of identifier

oandregal committed:
Add test for export redirects (* and default)

oandregal committed:
Placeholder to implement ExportAllDeclaration

oandregal committed:
Update engine and tests

oandregal committed:
Adapt get-name-declaration to return an array

oandregal committed:
Implement retrieval of multiple specifiers

oandregal committed:
Reorganize test to prepare next steps

oandregal committed:
Refactor default and namespace exports

oandregal committed:
Flatten importName and localName

Although the standard differentiate between the two
http://www.ecma-international.org/ecma-262/9.0/#exportentry-record
importName and localName are disjoint sets: if module is set,
the name is stored in importName variable and if it is unset,
the name is stored in localName variable.

Espree doesn't make that difference.

oandregal committed:
get-name-declaration: refactor to return export entry list

oandregal committed:
get-name-declaration renamed to get-export-entries

oandregal committed:
Reorder dependency statements

oandregal committed:
Remove uncessary test

oandregal committed:
Refactor get IR and add tests for default exports

oandregal committed:
Rename file

oandregal committed:
Flatten export entries returned by IR

oandregal committed:
Fix engine tests

oandregal committed:
Update tests for IR

oandregal committed:
Fix tests

oandregal committed:
getExportEntries: retrieve module from export statement

oandregal committed:
getIR: fix dependency lookup

oandregal committed:
Protect against no code provided

oandregal committed:
Better test names

oandregal committed:
Implement exported names

oandregal committed:
Fix exported names

oandregal committed:
Update coverage

oandregal committed:
Add test for namespace export

oandregal committed:
Update coverage.md

oandregal committed:
Fix export entries test

oandregal committed:
IR: add support for namespace exports

oandregal committed:
Fix method call

oandregal committed:
Extract constants for better encapsulation

oandregal committed:
Update coverage

oandregal committed:
Add test for named and default exports

oandregal committed:
Add JSDoc lookup for named inline nodes such as:

export function myFunction

oandregal committed:
Update coverage

oandregal committed:
Better test organization

oandregal committed:
Implement JSDoc lookup in default identifier from import

oandregal committed:
Better test organization

oandregal committed:
getExportEntries: test for _import default from module_

herregroen commented:
From what I can see doctrine is used here to parse the JSDoc. It seems to have a format that's at least similar to what JSDoc uses itself.

I've written https://github.com/herregroen/jsdoc-parser which functions as a JSDoc template and outputs to a json format that can be imported by https://github.com/WordPress/phpdoc-parser.

Ideally this would have an option to specify a JSON output format that would also produce docs that can be imported by the above plugin.

If that's the case then once https://meta.trac.wordpress.org/ticket/3063 gets resolved this could be used to generate the docs to be imported on developer.wordpress.org.

oandregal committed:
Add support for JSDoc lookup of imported identifiers

Example with default import:

	import myDefault from './module';
	...
	export { myDefault };

Example with named import:

	import { importedName as localName } from './module';
	...
	export { localName };

oandregal committed:
Update coverage

oandregal committed:
Install docgen package

oandregal committed:
Improve CLI

oandregal committed:
Add tests for retrieving JSDoc of destructured objects

oandregal committed:
Add support for JSDoc lookup of destructured objects

oandregal committed:
Adapt tests to new engine API

oandregal committed:
Fix JSDoc lookup for named export

oandregal committed:
Update coverage

oandregal committed:
Remove docgen from package.json

oandregal committed:
Better README

oandregal committed:
Update coverage

oandregal committed:
Ignore comments in namespace exports

oandregal committed:
Update coverage

oandregal committed:
CLI generates output by adding suffixes to input

oandregal committed:
Add test for export imported namespaces

oandregal committed:
Add test for export imported namespaces

oandregal committed:
Add named-import-namespace AST

oandregal committed:
Extract method

oandregal committed:
IR - add breaking test

oandregal committed:
Fix bug on export redirection

oandregal committed:
Add support for import namespace

oandregal committed:
Update coverage

oandregal committed:
Respect the dependency extension if any

oandregal committed:
Update coverage

oandregal committed:
Add test for undocumented declaration (no comments)

oandregal committed:
Add test for undocumented declaration (// comment)

oandregal committed:
Update README

oandregal committed:
Add a debug mode

oandregal committed:
Update coverage

oandregal committed:
Better example

oandregal committed:
Better README.md

oandregal committed:
Add params tags to IR and formatter

oandregal committed:
Formatter: show return tag output

oandregal committed:
Adapt engine tests to new API

oandregal committed:
Adapt formatter tests to new API

oandregal committed:
Adapt IR tests to new API

oandregal committed:
Update fixtures and test data to uncover some bug

mcsf commented:
> From what I can see doctrine is used here to parse the JSDoc. It seems to have a format that's at least similar to what JSDoc uses itself.
> 
> I've written https://github.com/herregroen/jsdoc-parser which functions as […]

@nosolosw and @herregroen, seems like we run the risk of duplicating efforts here. The way I see it, the atoms we need are:
- Extracting data from Gutenberg JS code. In particular, extract function signatures and JSDoc information.
- Produce this data in a language-agnostic format. Ideally, one that is already supported by what we use to generate pages on developer.wordpress.org from PHP code.
- Handle this kind of data more or less regardless of the language it refers to (JS or PHP).
- Run it through any kind of formatter (e.g., Markdown; some specific HTML template).

It appears we have loose pieces of work for roughly all of these, but they are spread out across properties and tickets and it's hard to keep track of them. Examples I've found:
- WordPress/phpdoc-parser#207: Add programming language support
- [meta#3063](https://meta.trac.wordpress.org/ticket/3063): Show JavaScript documentation on developer.wordpress.org
- [herregroen/jsdoc-parser](https://github.com/herregroen/jsdoc-parser/): Proposed JSDoc template
- WordPress/gutenberg#13329 (this very PR): docgen: JSDoc parser, Markdown formatter, etc.

Can we sync up on our needs and figure out how to converge our work? Let's make sure that the atoms can be worked on completely in parallel, which naturally entails agreeing on the dialects involved—to represent parsed data, etc.

oandregal committed:
Reorder tests

oandregal committed:
Fix bug on retrieving JSDoc from module dep

oandregal committed:
Markdown: clean spaces in descriptions

oandregal committed:
Add support for nullable types

oandregal committed:
Extract getParamType and add tests

oandregal committed:
getParamType: add support for RestType

oandregal committed:
getParamType: add support for unions

chrisvanpatten commented:
@nosolosw SUPER excited to see us consolidating efforts here! Love it!

mcsf commented:
Great to know! Let us know once you'd like a review.

oandregal committed:
Markdown tweaks: better formatting, adding output type

oandregal committed:
Markdown: sort output by symbol name

oandregal committed:
Markdown: improve spacing

oandregal committed:
Use | symbols for type unions

oandregal committed:
Use ... symbols for rest tipes

oandregal committed:
Markdown: update to consider uniont and rest types

oandregal committed:
getParamType: add support for nullable params

oandregal committed:
Markdown: add example tag to output

oandregal committed:
Update tests to API changes

oandregal committed:
Prevent bug when paragraph is null

oandregal committed:
Update coverage

oandregal committed:
Update coverage

oandregal committed:
Add support for see tag

oandregal committed:
Add support for link tag

oandregal committed:
Allow having several examples for one symbol

oandregal committed:
Less space between types: (object|string)

oandregal committed:
Keep description format as it is

oandregal committed:
Add support for type tag

oandregal committed:
Rename get-param-type-as-string to get-type-as-string

oandregal committed:
Adapt test to API changes

oandregal committed:
Add support for * type

oandregal committed:
Add support for type applications

oandregal committed:
Add support for NullLiteral

oandregal committed:
Add support for UndefinedLiteral and optional parameters

aduth commented:
A couple high-level thoughts:

- Doctrine was announced officially to be end-of-life ([reference](https://eslint.org/blog/2018/11/jsdoc-end-of-life))
- From your "TODO" notes of your `coverage.md` file, we should definitely not document `__unstable` APIs. I'd suggest even skipping `__experimental` APIs ([reference](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#experimental-apis)). Further, I think we might also take hints ([`@deprecated`](http://usejsdoc.org/tags-deprecated.html) JSDoc tag) to avoid documenting deprecated APIs, or otherwise clearly mark them as deprecated in the generated result.

nosolosw commented:
> Doctrine was announced officially to be end-of-life 

Ah, gotcha, thanks for the heads up @aduth. Will take look at alternatives.

> From your "TODO" notes of your `coverage.md` file, we should definitely not document `__unstable` APIs. I'd suggest even skipping `__experimental` APIs ([reference](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#experimental-apis)). Further, I think we might also take hints ([`@deprecated`](http://usejsdoc.org/tags-deprecated.html) JSDoc tag) to avoid documenting deprecated APIs, or otherwise clearly mark them as deprecated in the generated result.

Thanks for catching the `@deprecated` tag. In that case, I suggest adding a notice (for example, along the lines of [`UNSAFE_compomentWillMount`](https://reactjs.org/docs/react-component.html#unsafe_componentwillmount)).

For `__unstable`/`__experimental`/etc I was thinking about moving that decission to userland. My current thinking is to provide a way to tell the doc generator to omit certain symbol names.

oandregal committed:
Add support for default values

oandregal committed:
Add support for deprecated tag

oandregal committed:
params and return tags are no longer special

oandregal committed:
Refactor formatter to compact code

oandregal committed:
Update coverage

oandregal committed:
Add line numbers (start, end) to export tokens

oandregal committed:
Adapt tests to new API

oandregal committed:
Add line number to intermediate representation

oandregal committed:
Fix line numbers for namespace export

oandregal committed:
Remove duplicated tests

oandregal committed:
Better test names

oandregal committed:
Add path to intermediate representation

oandregal committed:
Adapt tests to API changes

oandregal committed:
Add path and line numbers to markdown output

oandregal committed:
Add TODO file

oandregal committed:
Use optionator to process CLI options

oandregal committed:
Allow to specify a markdown output file

oandregal committed:
Allow to pass a formatter

oandregal committed:
Document new added options

oandregal committed:
CLI: add ability to ignore symbols by name

oandregal committed:
Update README

oandregal committed:
CLI: support append to existing Markdown file

oandregal committed:
Make linter happy

oandregal committed:
Add package to manifest

oandregal committed:
Update as per new package recommendations

oandregal committed:
Update package name

oandregal committed:
Add docgen as a dev dependency

gziolo commented on packages/docgen/README.md:
```diff
@@ -0,0 +1,244 @@
+
+## Usage
+
+`node src/cli.js <entry-point.js>`
```
can it use bash syntax for consistency?

nosolosw replied:
399629a

gziolo commented on packages/docgen/README.md:
````diff
@@ -0,0 +1,244 @@
+
+Entry point:
+
+```js
````
Out of curiosity, do you use the tool to generate the output of those examples?
If not, maybe it would make sense to find a way to do it, format can change so it would be easier to keep it in sync :)

nosolosw replied:
Yeah, I manually created some examples and then generated this output. I agree a more streamlined process would be great.

Perhaps a potential way to improve this would be what you proposed at https://github.com/WordPress/gutenberg/pull/13856#discussion_r256843355 If we had a way to mark a Markdown file with tokens (instead of the current system that uses a section title) it'd be useful for this as well.

nosolosw replied:
I've implemented a new CLI option `--to-token` that allows inserting the autogenerated docs within the Markdown file.

gziolo commented on packages/docgen/package.json:
```diff
@@ -0,0 +1,44 @@
+	"publishConfig": {
+		"access": "public"
+	},
+	"scripts": {
```
This is different than all other packages. Unit tests are executed from the root folder for everything else. It should be documented for future maintainers how to run those tests. You also need to keep in mind that to install `devDependencies` in here someone will have to explicitly go to this folder. It won't get installed with `npm i` in the root folder. Only `dependencies` get hoisted.

nosolosw replied:
Happy to adapt this to the way other packages are set up. I'm just unfamiliar with the lerna monorepo setup and haven't found any doc about how should we create new packages. Do you have any reference I can check? Or specific steps?

gziolo replied:
Technically speaking we can leave it as is. I just wanted to make sure you are aware of the implications of such setup.

For reference, this is how you can run a script from the package maintained by Lerna:
https://github.com/WordPress/gutenberg/blob/master/package.json#L154
`lerna run build` in particular is the way to run all `build` scripts in packages. See more at https://github.com/lerna/lerna/tree/master/commands/run

nosolosw replied:
Fixed what you mentioned:

- tests are now migrated to jest, so they're picked by it and executed as any other WordPress package.
- dependencies are now hoisted.

gziolo commented on packages/docgen/package.json:
```diff
@@ -0,0 +1,44 @@
+	"author": "The WordPress Contributors",
+	"license": "GPL-2.0-or-later",
+	"keywords": [
+		"jsdoc",
```
We always put `wordpress` keyword in this array.

nosolosw replied:
b08b8d3cf3824addaac12c546316c909e7a38194

oandregal committed:
Add wordpress keyword to package.json

oandregal committed:
Mark as a bash command

nosolosw commented on packages/docgen/README.md:
````diff
@@ -0,0 +1,246 @@
+
+## Usage
+
+```bash
````
Rebasing from master seems to have deleted your previous comment, @gziolo. For reference 399629ad2ae8b067aa85f7174d77a8184fa4c9f7 added `bash` syntax to this as well.

oandregal committed:
CLI: allow for walking parent directories

nosolosw commented:
Did some research on doctrine alternatives (will update if needed):

Library | Latest publish (months ago) | npm weekly downloads (thousands) | Comments
--- | --- | --- | --- 
eslint-plugin-jsdoc | 1 | [30](https://www.npmjs.com/package/eslint-plugin-jsdoc) | Uses comment-parser and jsdoctypeparser.
comment-parser | 1 | [120](https://www.npmjs.com/package/comment-parser) | Parses JSDoc comment, but not types.
jsdoctypeparser | 36 | [100](https://www.npmjs.com/package/jsdoctypeparser) | Google Closure and JSDoc type parser.
jsdoc | 12 | [235](https://www.npmjs.com/package/jsdoc) | Uses catharsis for type parser and their own thing for the rest.
catharsis | 24 | [224](https://www.npmjs.com/package/catharsis) | Google Closure and JSDoc type parser.
doctrine | 3 | [8.000](https://www.npmjs.com/package/doctrine) | Developed their own parsers.

oandregal committed:
Add tests for tags

oandregal committed:
Adapt formatter test to API

oandregal committed:
Move IR tests related to JSDoc to its own tester

oandregal committed:
get-jsdoc-from-token returns a human-readable type

nosolosw commented:
At this point, I think it makes sense to keep `doctrine` until a better alternative is developed.

I don't see a mature alternative to `doctrine` yet: I haven't found how to use `jsdoc` as a library to parse a JSDoc comment, and other libraries such as `comment-parser` have some gotchas (such as: it isn't able to [parse tags without a name out of the box](https://github.com/yavorskiy/comment-parser/issues/21) -`return`, `link`, `see`, `deprecated`- although it is fixable by writing custom parser for each tag.

In 5b3fcb7, b8ba3d0, 1bde19e, and 1bdcb95 I've added some tests and wrapped everything related to `doctrine` within the [`get-jsdoc-from-token`](https://github.com/WordPress/gutenberg/blob/1bdcb9592b98921128c82f6e2a97716f678b6878/packages/docgen/src/get-jsdoc-from-token.js) module, so it's easy to switch without affecting the `docgen` public API when time comes.

oandregal committed:
Implement a token-based injection for Markdown

oandregal committed:
Add embed JSDoc

oandregal committed:
Formatter: let heading index be configurable

oandregal committed:
Use proper heading when embedding content within token

oandregal committed:
Move default markdown code to own folder

oandregal committed:
Migrate engine tests to jest

oandregal committed:
Reorganize fixtures

oandregal committed:
Migrate export tests to jest

oandregal committed:
Migrate getTypeFromJSDoc tests to jest

oandregal committed:
Migrate getJSDocFromToken tests to jest

oandregal committed:
Migrate getIntermediateRepresentation tests to jest

oandregal committed:
Migrate Markdown formatter tests to jest

oandregal committed:
Extract bin/cli utility

oandregal committed:
Uninstall no longer needed dev deps

oandregal committed:
Fix docgen dependencies

oandregal committed:
Add command to generate docs

- Add command for e2e-test-utils
- Add command in top-level package.json as to simplify executing
  across packages without lerna knowledge.

oandregal committed:
Update docs

oandregal committed:
Update README

oandregal committed:
Add --use-token option

oandregal committed:
Extract constants

oandregal committed:
Update README

nosolosw commented:
I've addressed Greg comments and this is ready for a second round of review.

cc @herregroen @mcsf @chrisvanpatten @aduth and any other person interested :)

oandregal committed:
Make formatter-markdown test be deterministic

oandregal committed:
Make formatter-markdown test be deterministic

chrisvanpatten commented on packages/docgen/README.md:
````diff
@@ -0,0 +1,250 @@
+npx docgen <entry-point.js>
+```
+
+This command will generate a file named `entry-point-api.md` containing all the exports and its JSDoc comments.
````
```suggestion
This command will generate a file named `entry-point-api.md` containing all the exports and their JSDoc comments.
```

chrisvanpatten commented on packages/docgen/README.md:
```diff
@@ -0,0 +1,250 @@
+* **--to-token**: Embed generated documentation within the start and end tokens in the Markdown output. To be used by the default Markdown formatter.Depends on `--output` and bypasses the custom `--formatter` passed, if any.
+  * Start token: `<!-- START TOKEN(Autogenerated API docs) -->`
+  * End token: `<!-- END TOKEN(Autogenerated API docs) -->`
+* **--use-token** `(String)`: This options allows to customize the string between the tokens. For example, `--use-token my-api` will look up for the start token `<!-- START TOKEN(my-api) -->` and the end token `<!-- END TOKEN(my-api) -->`. Depends on `--to-token`.
```
```suggestion
* **--use-token** `(String)`: This options allows you to customize the string between the tokens. For example, `--use-token my-api` will look up for the start token `<!-- START TOKEN(my-api) -->` and the end token `<!-- END TOKEN(my-api) -->`. Depends on `--to-token`.
```

chrisvanpatten commented on packages/docgen/README.md:
```diff
@@ -0,0 +1,250 @@
+  * Start token: `<!-- START TOKEN(Autogenerated API docs) -->`
+  * End token: `<!-- END TOKEN(Autogenerated API docs) -->`
+* **--use-token** `(String)`: This options allows to customize the string between the tokens. For example, `--use-token my-api` will look up for the start token `<!-- START TOKEN(my-api) -->` and the end token `<!-- END TOKEN(my-api) -->`. Depends on `--to-token`.
+* **--debug**: Run in debug mode, which outputs some intermediate files useful for debugging
```
```suggestion
* **--debug**: Run in debug mode, which outputs some intermediate files useful for debugging.
```

chrisvanpatten commented on packages/docgen/README.md:
```diff
@@ -0,0 +1,250 @@
+
+Some characteristics:
+
+* If the export statement doesn't contain any JSDoc, it'll look up for JSDoc up to the declaration.
```
I'm not sure I understand this. Can you clarify?

chrisvanpatten replied:
I guess you're saying if there's no JSDoc on the export, it will look for the JSDoc with the exported thing itself?

nosolosw replied:
Exactly. For example:

```js
/**
 * Adds two numbers.
 *
 * @param {number} term1 First number.
 * @param {number} term2 Second number.
 * @return {number} The result of adding the two numbers.
 */
function addition( term1, term2 ) {
	// Implementation would go here.
}

export default addition;
```

The export statement doesn't have any JSDoc, but the symbol declaration does.

mkaz reviewed (approved):
Great work! 👍 This works nicely as described. I tested it on a few of the existing ES6 packages and it worked as expected. The next piece will be how to hook and generate for existing documentation but that will be another PR.

One small suggestion, but wouldn't consider it a blocker since you do have the example command above, is adding in the examples the file names and command used, for example:

Entry point: `index.js`

```
/**
 * Adds two numbers.
 *
 * @param {number} term1 First number.
 * @param {number} term2 Second number.
 * @return {number} The result of adding the two numbers.
 */
export default function addition( term1, term2 ) {
	// Implementation would go here.
}
```
Command: `npx docgen index.js`

Output: `index-api.md`

```
# API

## default

[example.js#L8-L10](example.js#L8-L10)

Adds two numbers.

**Parameters**

- **term1** `number`: First number.
- **term2** `number`: Second number.

**Returns**

`number` The result of adding the two numbers.
```

chrisvanpatten committed:
Update packages/docgen/README.md

Co-Authored-By: nosolosw <nosolosw@users.noreply.github.com>

chrisvanpatten committed:
Update packages/docgen/README.md

Co-Authored-By: nosolosw <nosolosw@users.noreply.github.com>

chrisvanpatten committed:
Update packages/docgen/README.md

Co-Authored-By: nosolosw <nosolosw@users.noreply.github.com>

oandregal committed:
Add command in the examples

nosolosw commented:
> The next piece will be how to hook and generate for existing documentation but that will be another PR.

I was tempted to do it here as well, but it'd add unnecessary noise. I actually added the command for the `e2e-test-utils` package, though, just because it was the first testbed for this PR and I wanted to demonstrate how it would fit.

> One small suggestion, but wouldn't consider it a blocker since you do have the example command above, is adding in the examples the file names and command used

Nice suggestion. Implemented at 0f364a2.

oandregal committed:
Update package-lock using the latest npm

nosolosw commented:
This was merged everyone, let's get the docs :tada: started!

aduth commented on packages/docgen/src/engine.js:
```diff
@@ -0,0 +1,56 @@
+/**
```
I wonder if we should have avoided placing these files in a directory called `src/`, because for every other package we use the `src` directory to indicate pre-transpiled code, with the expectation there'd be a complementing `build` directory as output. I encountered this today when auditing our transpiled packages for `@babel/runtime` and became confused when this package was not respecting the convention.

In other words, I see no reason why all of these files couldn't just live in the root of the package, or maybe we could create a different convention for organizing files within packages which don't need transpiling.

aduth replied:
In fact, `src/` is used as a trigger to process files via Babel.

You can see that `docgen` includes `build` and `build-module` outputs, even though they're effectively dead code:

https://unpkg.com/@wordpress/docgen@1.0.0/

Related code:

https://github.com/WordPress/gutenberg/blob/9c3dbde9f7585ed776fa52a3720ad581d556f865/bin/packages/build.js#L193-L200

gziolo replied:
We can change this logic and decide whether a given package should be transpiled based on the existance of `module` entry in `package.json` file.

gziolo replied:
I see what you mean. `docgen` never uses transpiled code but it's still published to npm.

nosolosw replied:
> In fact, src/ is used as a trigger to process files via Babel.

Didn't know that until now.

> We can change this logic and decide whether a given package should be transpiled based on the existance of module entry in package.json file.

I like this. It allows us to store code within `src/` for complex packages like `docgen` that require multiple files. It's also in line with the idea we discussed somewhere about favoring `package.json` config over maintaining ignore lists.

aduth commented on packages/docgen/src/index.js:
```diff
@@ -0,0 +1,126 @@
+		currentFileStack.push( inputFile );
+		const relativePath = path.relative( rootDir, inputFile );
+		const result = engine( relativePath, data, getIRFromRelativePath( rootDir, last( currentFileStack ) ) );
+		currentFileStack.pop( inputFile );
```
`Array#pop` takes no arguments. It could be safely removed.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/pop

nosolosw replied:
Mind to add this in the `@private` PR?

aduth replied:
> Mind to add this in the `@private` PR?

I made a small pull request: #15397

aduth commented on packages/docgen/src/get-export-entries.js:
```diff
@@ -0,0 +1,97 @@
+				case 'AssignmentExpression':
+					name = t.declaration.left.name;
+					break;
+				//case 'FunctionDeclaration'
```
Should have been removed?

nosolosw replied:
From memory, I think it documents the cases that fall-through.

aduth replied:
> From memory, I think it documents the cases that fall-through.

It could also be explicit, no?

```js
switch ( t.declaration.type ) {
	// ...

	case 'FunctionDeclaration':
	case 'ClassDeclaration':
	default:
		// ...
}
```

_(If it were a web bundle, we may have concern about the "unnecessary" bytes added, but it's not in this instance)_

I'd just worry that this might not be obvious to the future maintainer without at least a hint (inline comment) to explain the intention.

aduth commented on packages/docgen/src/test/fixtures/default-class-named/exports.json:
```diff
@@ -0,0 +1,101 @@
+{
```
How are these files generated?

nosolosw replied:
Using docgen's `--debug` [mode](https://github.com/WordPress/gutenberg/blob/master/packages/docgen/README.md#cli-options). 

From the root of the repo:

>npx docgen --debug packages/docgen/src/test/fixtures/default-class-named/code.js

nosolosw replied:
I seem to have tweaked the exports files removing the enclosing `[]`. In retrospect, don't know why I thought that was a good idea.
